src/box/field_default_func.c

```diff
@@ -0,0 +1,134 @@
+/*
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright 2010-2023, Tarantool AUTHORS, please see AUTHORS file.
+ */
+#include "port.h"
+#include "engine.h"
+#include "tt_static.h"
+#include "tuple_format.h"
+#include "field_default_func.h"
+
+/**
+ * Find `default_func` in func cache.
+ * Return NULL if something went wrong (diag is set).
+ */
+struct func *
+field_default_func_find(struct field_default_func *default_func)
+{
+	struct func *func = func_by_id(default_func->id);
+	if (func == NULL) {
+		diag_set(ClientError, ER_NO_SUCH_FUNCTION,
+			 int2str(default_func->id));
+	}
+	return func;
+}
+
+/**
+ * Check `func` for compliance with the field default function rules.
+ *
+ * Returns 0 on success. On error, sets diag and returns -1.
+ * `fieldno` is used for better error message.
+ */
+static int
+field_default_func_verify(struct func *func, uint32_t fieldno)
+{
+	const char *func_name = func->def->name;
+
+	if (func->def->language == FUNC_LANGUAGE_LUA &&
+	    func->def->body == NULL) {
+		diag_set(ClientError, ER_WRONG_SPACE_FORMAT,
+			 fieldno + TUPLE_INDEX_BASE,
+			 tt_sprintf("field default Lua function '%s' must have "
+				    "persistent body", func_name));
+		return -1;
+	}
+	return 0;
+}
+
+/**
+ * Function that calls field default function `default_func`.
+ * `arg_data` points to msgpack data with the function argument.
+ * `ret_data` is pointed to msgpack data with the function return value,
+ * allocated on current fiber's region.
+ *
+ * Returns 0 on success. On error, sets diag and returns -1.
+ */
+static int
+field_default_func_call(struct field_default_func *default_func,
+			const char *arg_data, uint32_t arg_size,
+			const char **ret_data, uint32_t *ret_size)
+{
+	struct port out_port, in_port;
+	port_c_create(&in_port);
+	if (arg_data != NULL)
+		port_c_add_mp(&in_port, arg_data, arg_data + arg_size);
+
+	int rc = func_call_no_access_check(default_func->holder.func,
+					   &in_port, &out_port);
+	port_destroy(&in_port);
+	if (rc != 0) {
+		diag_log();
+		diag_clear(diag_get());
+		diag_set(ClientError, ER_FIELD_DEFAULT_FUNC,
```

Comment:
Nit: The error message format looks confusing - better add a colon before the reason, for example: `Error calling default value function '%s': %s`. Also, I think it's better to include the original error message after a colon instead of logging it (use `diag_last_error()->errmsg`).

Reply:
Done.